codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/SerializeMiddleware.java

```diff
@@ -0,0 +1,165 @@
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.utils.MapUtils;
+
+public class SerializeMiddleware {
+    protected final ProtocolGenerator.GenerationContext ctx;
+    protected final OperationShape operation;
+    protected final GoWriter writer;
+
+    protected final StructureShape input;
+    protected final StructureShape output;
+
+    private String serialName;
+
+    public SerializeMiddleware(
+            ProtocolGenerator.GenerationContext ctx, OperationShape operation, GoWriter writer
+    ) {
+        this.ctx = ctx;
+        this.operation = operation;
+        this.writer = writer;
+
+        this.input = ctx.getModel().expectShape(operation.getInputShape(), StructureShape.class);
+        this.output = ctx.getModel().expectShape(operation.getOutputShape(), StructureShape.class);
+    }
+
+    public GoWriter.Writable generate() {
+        serialName = "awsAwsjson10_serializeOp" + operation.toShapeId().getName();
+
+        return goTemplate("""
+
+                type $opName:L struct{
+                }
+
+                func (op *$opName:L) ID() string {
+                    return "OperationSerializer"
+                }
+
+                $handleSerialize:W
+
+                """,
+                MapUtils.of(
+                    "opName", serialName,
+                    "handleSerialize", generateHandleSerialize()
+        ));
+    }
+
+    private GoWriter.Writable generateHandleSerialize() {
+        return goTemplate(
+                """
+
+                        func (op *$opName:L) HandleSerialize (ctx $context:T, in $input:T, next $handler:T) (
+                        out $output:T, metadata $metadata:T, err error) {
+
+                            $body:W
+
+                            return next.HandleSerialize(ctx, in)
+                        }
+
+                        """, MapUtils.of(
+                        "context", SmithyGoDependency.CONTEXT.interfaceSymbol("Context"),
+                        "input", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeInput"),
+                        "handler", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeHandler"),
+                        "output", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeOutput"),
+                        "metadata", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("Metadata"),
+                        "opName", serialName,
+                        "body", generateHandleSerializeBody())
+        );
+    }
+
+    private GoWriter.Writable generateHandleSerializeBody() {
+        return goTemplate("""
+                    $route:W
+
+                    $serialize:W
+
+                """,
+                MapUtils.of(
+                        "route", handleHttp(),
+                        "serialize", handlePayload()
+                ));
+    }
+
+    private GoWriter.Writable handleHttp() {
```

Comment:
This name doesn't really make sense or tell us anything.

Reply:
Would handleHttpRequest be better? The function error checks the input parameters and request and then sets up the header for the HTTP request for the respective portion of the "Handle Serialize" function - I'm not sure what else to name it.

Reply:
That's not really any different, just more verbose.

You've said there what it's doing, just translate that to code. I wouldn't think too hard about it.

"handleProtocolSetup"

"handleRequestPreamble"

"generateRequestPreamble"

"generateProtocolHeaders"

etc.

Reply:
Ah got it def was overthinking it.